python/src/opendp/polars.py

```diff
@@ -190,3 +196,180 @@ def lazy(self):
+            if callable(attr):
+
+                def wrap(*args, **kwargs):
+                    out = attr(*args, **kwargs)
+
+                    # We keep the query plan void of data anyways,
+                    # so running the computation doesn't affect privacy.
+                    # This isn't meant to cover all possible APIs that may execute the query,
+                    #    but it does give a simple sanity check in the obvious cases.
+                    if isinstance(out, DataFrame):
+                        raise ValueError(
+                            "You must call `.release()` before executing a query."
+                        )
+
+                    # re-wrap any lazy outputs to keep the conveniences afforded by this class
+                    if isinstance(out, (LazyGroupBy, LazyFrame)):
+                        return LazyFrameQuery(out, query)
+
+                    return out
+
+                return wrap
+            return attr
+
+        def resolve(self) -> Measurement:
+            """Resolve the query into a measurement."""
+            from opendp.context import PartialChain
+
+            # access attributes of self without getting intercepted by Self.__getattribute__
+            lf_plan = object.__getattribute__(self, "_lf_plan")
+            query = object.__getattribute__(self, "_query")
+            input_domain, input_metric = query._chain
+
+            # create an updated query and then resolve it into a measurement
+            return query.new_with(
+                chain=PartialChain(
+                    lambda s: make_private_lazyframe(
+                        input_domain=input_domain,
+                        input_metric=input_metric,
+                        output_measure=query._output_measure,
+                        lazyframe=lf_plan,
+                        global_scale=s,
+                    )
+                )
+            ).resolve()
+
+        def release(self) -> OnceFrame:
+            """Release the query. The query must be part of a context."""
+            query = object.__getattribute__(self, "_query")
+            resolve = object.__getattribute__(self, "resolve")
+            return query._context(resolve())  # type: ignore[misc]
+
+except ImportError:
+    ERR_MSG = (
+        "LazyFrameQuery needs Polars to be installed: `pip install 'opendp[polars]'`."
+    )
+
+    class LazyFrameQuery(object):  # type: ignore[no-redef]
+        """LazyFrameQuery mimics a Polars LazyFrame, but makes a few additions and changes as documented below."""
+
+        def resolve(self) -> Measurement:
+            """Resolve the query into a measurement."""
+            raise ImportError(ERR_MSG)
+
+        def release(self) -> OnceFrame:
+            """Release the query. The query must be part of a context."""
+            raise ImportError(ERR_MSG)
+
+
+@dataclass
+class Margin(object):
+    public_info: Literal["keys"] | Literal["lengths"] | None = None
```

Comment:
huh: I was thinking this `|` syntax wasn't supported in 3.9, and I thought I saw a CI error earlier. Was I confused then, or is mypy not checking this?

Reply:
It is supported if you use `from __future__ import annotations`. However you still have to be careful, as this can cause some weird edge cases with Sphinx if you use it in conjunction with TypeVars.

Reply:
Cool: I'll file a follow-up issue to move to this syntax across the codebase, so we're not deciding on a case-by-case basis which syntax to use.